components/openmpi-controller/controller/controller.py

```diff
@@ -43,10 +50,20 @@ def __exit__(self, exc_type, exc_val, exc_tb):
     def wait_ready(self):
         if self.num_gpus > 0:
             self._wait_nvidia_driver_present()
+        self._download_data()
         Path(SIGCONT).touch()
 
     def wait_done(self):
         self._wait_master_terminated()
+        self._upload_data()
```

Comment:
I think we should make this method and `download_data()` public and call it explicitly in `main.py`.  Maybe rename to `upload_data_if_needed()`, `download_data_if_needed()`

Reply:
Currently wait_ready creates SIGCONT file in wait_ready. If we have to call download_data_if_needed in main, then it must be called before wait_ready, which seems to expose too much information to to caller.

Later, we may want to do "s3 sync" periodically instead of uploading data at the end. I think it makes more sense to keep all the data sync logic within the controller.

Reply:
OK, that sounds reasonable.